src/libraries/System.Private.CoreLib/src/System/Reflection/FieldAccessor.cs

```diff
@@ -0,0 +1,518 @@
+                switch (_fieldAccessType)
+                {
+                    case FieldAccessorType.InstanceReferenceType:
+                        VerifyTarget(obj);
+                        Debug.Assert(obj != null);
+                        return Volatile.Read(ref Unsafe.As<byte, object>(ref Unsafe.AddByteOffset(ref obj.GetRawData(), _addressOrOffset)));
+
+                    case FieldAccessorType.InstanceValueType:
+                    case FieldAccessorType.InstanceValueTypeSize1:
+                    case FieldAccessorType.InstanceValueTypeSize2:
+                    case FieldAccessorType.InstanceValueTypeSize4:
+                    case FieldAccessorType.InstanceValueTypeSize8:
+                        VerifyTarget(obj);
+                        Debug.Assert(obj != null);
+                        return RuntimeHelpers.Box(
+                            _methodTable,
+                            ref Unsafe.AddByteOffset(ref obj.GetRawData(), _addressOrOffset));
+
+                    case FieldAccessorType.InstancePointerType:
+                        VerifyTarget(obj);
+                        Debug.Assert(obj != null);
+                        return Pointer.Box(
+                            (void*)Unsafe.As<byte, IntPtr>(ref Unsafe.AddByteOffset(ref obj.GetRawData(), _addressOrOffset)),
+                            _fieldInfo.FieldType);
+
+                    case FieldAccessorType.StaticReferenceType:
+                        return Volatile.Read(ref Unsafe.As<IntPtr, object>(ref *(IntPtr*)_addressOrOffset));
+
+                    case FieldAccessorType.StaticValueType:
+                    case FieldAccessorType.StaticValueTypeSize1:
+                    case FieldAccessorType.StaticValueTypeSize2:
+                    case FieldAccessorType.StaticValueTypeSize4:
+                    case FieldAccessorType.StaticValueTypeSize8:
+                        return RuntimeHelpers.Box(_methodTable, ref Unsafe.AsRef<byte>(_addressOrOffset.ToPointer()));
+
+                    case FieldAccessorType.StaticValueTypeBoxed:
+                        // Re-box the value.
+                        return RuntimeHelpers.Box(
+                            _methodTable,
+                            ref Unsafe.As<IntPtr, object>(ref *(IntPtr*)_addressOrOffset).GetRawData());
+
+                    case FieldAccessorType.StaticPointerType:
+                        return Pointer.Box((void*)Unsafe.As<byte, IntPtr>(
+                            ref Unsafe.AsRef<byte>(_addressOrOffset.ToPointer())), _fieldInfo.FieldType);
+
+                    case FieldAccessorType.SlowPathUntilClassInitialized:
+                        if (!IsStatic())
+                        {
+                            VerifyTarget(obj);
+                        }
+
+                        object? ret = RuntimeFieldHandle.GetValue(_fieldInfo, obj, (RuntimeType)_fieldInfo.FieldType, _fieldInfo.m_declaringType, isInitialized: false);
+                        Initialize();
```

Comment:
This is not right. We cannot assume that the type is initialized when we fetch the field successfully.

Consider this example:
```
using System;
using System.Reflection;

Test();
for (int j = 0; j < 10; j++) GC.Collect();
Test();

static void Test()
{
    FieldInfo fi = typeof(MyType).GetField("f");
    for (int i = 0; i < 3; i++)
    {
        bool exceptionThrown = false;
        try
        {
            fi.GetValue(null);
        }
        catch (Exception e)
        {
            Console.WriteLine(e);
            exceptionThrown = true;
        }
        if (!exceptionThrown)
        {
            Console.WriteLine("No exception!");
        }
    }
}

public class MyType
{
    public static int f;

    static MyType()
    {
        FieldInfo fi = typeof(MyType).GetField("f");
        for (int i = 0; i < 3; i++)
            fi.GetValue(null);
        throw new Exception();
    }
}
```

The first GetField call is going to throw `System.TypeInitializationException:` (correct). The next couple of GetField calls are going to succeed (incorrect), ignoring the fact that the type is not yet initialized. Then we force GCs to get the field infos re-created and the GetField calls are going throw `System.TypeInitializationException` again.

This was broken in a slightly different way even before your change.

Either way, we need the VM to explicitly tell us that the type is initialized before switching to fast path.

Reply:
(Convert my example to a test.)

Reply:
I see - the runtime only allows the cctor to be called once, and then caches the exception. Added the test which did fail before any changes made here.